controllers/search_controller.go

```diff
@@ -230,6 +223,25 @@ func (r *SearchReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctr
 	once.Do(func() {
 		addon.CreateAddonOnce(ctx, instance)
 	})
+	cleanUpFunc := func() {
+		// delete legacy servicemonitor setup
+		result, err = r.deleteLegacyServiceMonitorSetup(instance)
+		cleanupComplete = true
+		log.Info("cleanupComplete. Setting cleanupComplete to ", "cleanupComplete", cleanupComplete)
+		if result != nil {
+			cleanupComplete = false
+			log.Info("cleanup not complete. Setting cleanupComplete to ", "cleanupComplete", cleanupComplete)
+			log.Error(err, "Failed to remove legacy ServiceMonitor setup ")
+		}
+	}
+	if !cleanupComplete {
+		cleanOnce.Do(cleanUpFunc)
```

Comment:
`cleanupComplete` ensures that we cleanup once or retry if failed.
So then, is  `cleanOnce.Do(...)` redundant here?

Reply:
Does it retry on every reconcile?  I wonder if an unrecoverable error will continue to happen in a loop and create problems here.

Reply:
It retries every loop. Rewrote that part so that it doesn't retry. A pod restart will fix in case any artifact remains.